pkg/addons/default/kube_proxy.go

```diff
@@ -20,55 +20,55 @@ const (
 )
 
 // UpdateKubeProxyImageTag updates image tag for kube-system:damoneset/kube-proxy based to match controlPlaneVersion
-func UpdateKubeProxyImageTag(clientSet kubernetes.Interface, controlPlaneVersion string, dryRun bool) error {
+func UpdateKubeProxyImageTag(clientSet kubernetes.Interface, controlPlaneVersion string, plan bool) (bool, error) {
 	printer := printers.NewJSONPrinter()
 
 	d, err := clientSet.AppsV1().DaemonSets(metav1.NamespaceSystem).Get(KubeProxy, metav1.GetOptions{})
 	if err != nil {
 		if apierrs.IsNotFound(err) {
 			logger.Warning("%q was not found", KubeProxy)
-			return nil
+			return false, nil
 		}
-		return errors.Wrapf(err, "getting %q", KubeProxy)
+		return false, errors.Wrapf(err, "getting %q", KubeProxy)
 	}
 	if numContainers := len(d.Spec.Template.Spec.Containers); !(numContainers >= 1) {
-		return fmt.Errorf("%s has %d containers, expected at least 1", KubeProxy, numContainers)
+		return false, fmt.Errorf("%s has %d containers, expected at least 1", KubeProxy, numContainers)
 	}
 
 	if err := printer.LogObj(logger.Debug, KubeProxy+" [current] = \\\n%s\n", d); err != nil {
-		return err
+		return false, err
 	}
 
 	image := &d.Spec.Template.Spec.Containers[0].Image
 	imageParts := strings.Split(*image, ":")
 
 	if len(imageParts) != 2 {
-		return fmt.Errorf("unexpected image format %q for %q", *image, KubeProxy)
+		return false, fmt.Errorf("unexpected image format %q for %q", *image, KubeProxy)
 	}
 
 	desiredTag := "v" + controlPlaneVersion
 
 	if imageParts[1] == desiredTag {
 		logger.Debug("imageParts = %v, desiredTag = %s", imageParts, desiredTag)
 		logger.Info("%q is already up-to-date", KubeProxy)
-		return nil
+		return false, nil
```

Comment:
Seems like the return value is completely dependent of the input argument `plan`. I would remove the return boolean. It's a bit confusing.

Reply:
Yes, almost.

The idea is to print this
```
foo already up-to-date
```

vs this:
```
foo needs an update
re-run with '--approve'
```
A simple solution would be to print that line right away here, but that will break API/CLI boundary. So the idea is that the logic that decides whether to print `re-run with '--approve'` or not must be outside of this function. returning a boolean value may not be ideal, but I didn't want to over-engineer this part.
There is a boolean `plan` argument, and a boolean return value on whether approved plan would make any changes or not, that seems plausible to me. If error value matching was more flexible, perhaps we could have used the error value somehow, but Go is unhelpful with that.